src/Bridge/LibraryBridgeHelper.cpp

```diff
@@ -61,26 +70,91 @@ void LibraryBridgeHelper::startBridge(std::unique_ptr<ShellCommand> cmd) const
+    if (!parsed || (dictionary_id_exists != 0 && dictionary_id_exists != 1))
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected message from library bridge: {} ({}). Check bridge and server have the same version.",
+                        result, parsed ? toString(dictionary_id_exists) : "failed to parse");
+
+    LOG_TRACE(log, "dictionary_id: {}, dictionary_id_exists on bridge side: {}, library confirmed to be initialized on server side: {}",
+              toString(dictionary_id), toString(dictionary_id_exists), library_initialized);
+
+    if (dictionary_id_exists && !library_initialized)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Library was not initialized, but bridge responded to already have dictionary id: {}", dictionary_id);
+
+    if (!dictionary_id_exists && library_initialized)
+    {
+        LOG_WARNING(log, "Library bridge does not have library handler with dictionaty id: {}. It will be reinitialized.", dictionary_id);
+        bool reinitialized = false;
+        try
+        {
+            reinitialized = initLibrary(false);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log);
+            return false;
+        }
+
+        if (!reinitialized)
+            throw Exception(ErrorCodes::EXTERNAL_LIBRARY_ERROR,
+                            "Failed to reinitialize library handler on bridge side for dictionary with id: {}", dictionary_id);
+    }
+
+    return true;
+}
 
+
+ReadWriteBufferFromHTTP::OutStreamCallback LibraryBridgeHelper::getInitLibraryCallback() const
+{
     /// Sample block must contain null values
     WriteBufferFromOwnString out;
     auto output_stream = getContext()->getOutputStream(LibraryBridgeHelper::DEFAULT_FORMAT, out, sample_block);
     formatBlock(output_stream, sample_block);
     auto block_string = out.str();
 
-    auto out_stream_callback = [library_path, library_settings, attributes_names, block_string, this](std::ostream & os)
+    return [block_string, this](std::ostream & os)
     {
-        os << "library_path=" << escapeForFileName(library_path) << "&";
-        os << "library_settings=" << escapeForFileName(library_settings) << "&";
-        os << "attributes_names=" << escapeForFileName(attributes_names) << "&";
+        os << "library_path=" << escapeForFileName(library_data.library_path) << "&";
+        os << "library_settings=" << escapeForFileName(library_data.library_settings) << "&";
+        os << "attributes_names=" << escapeForFileName(library_data.dict_attributes) << "&";
         os << "sample_block=" << escapeForFileName(sample_block.getNamesAndTypesList().toString()) << "&";
         os << "null_values=" << escapeForFileName(block_string);
     };
-    return executeRequest(uri, out_stream_callback);
+}
+
+
+bool LibraryBridgeHelper::initLibrary(bool check_bridge) const
+{
+    /// Do not check if we call initLibrary from checkBridgeSync.
+    if (check_bridge)
+        startBridgeSync();
```

Comment:
`initLibrary -> startBirdgeSync -> checkBridgeIsRunning -possibly> initLibrary ->` .... 

I'm confused. Can we do it simpler?

Reply:
Actually such branch
`initLibrary -> startBirdgeSync -> checkBridgeIsRunning -possibly> initLibrary ->`
is not possible because of `is_initialized`, but yes, need to make it in a better way.